### PR TITLE
UnivariatePolynomial and UnivariateExprPolynomial

### PR DESCRIPTION
**Pull request by chenchfort**
A PR for UnivariatePolynomial (only supports integer coefficients for now) and UnivariateExprPolynomial that overloads operators for UnivariatePolynomial.

### CODE REVIEW

File: symengine/expand.cpp [hunk 1]
```diff
@@ -239,49 +239,93 @@ class ExpandVisitor : public BaseVisitor<ExpandVisitor> {
...
-            int q = rcp_static_cast<const Integer>(self.get_exp())->as_int();
-            RCP<const UnivariateIntPolynomial> p = rcp_static_cast<const UnivariateIntPolynomial>(_base);
-            RCP<const UnivariateIntPolynomial> r = univariate_int_polynomial(p->var_, 0, {{0, integer_class(1)}});
-            while (q != 0) {
-                if (q % 2 == 1) {
-                    r = mul_poly(r, p);
-                    q--;
+        if(is_a<const UnivariatePolynomial>(*self.get_base())) {
+            if (is_a<Integer>(*self.get_exp()) && is_a<UnivariatePolynomial>(*_base)) {
+                int q = rcp_static_cast<const Integer>(self.get_exp())->as_int();
```
Comment (isuruf) [inline symengine/expand.cpp]: Can you implement this function as a separate function in `ExpandVisitor` and call it from here?

File: symengine/polynomial.cpp [hunk 2]
```diff
@@ -8,24 +8,21 @@ namespace SymEngine {
 
 UnivariateIntPolynomial::UnivariateIntPolynomial(const RCP<const Symbol> &var, const unsigned int &degree, map_uint_mpz&& dict) :
      degree_{degree}, var_{var}, dict_{std::move(dict)} {
-
```
Comment (isuruf) [inline symengine/polynomial.cpp]: This change is unnecessary

Comment (chenchfort) [inline symengine/polynomial.cpp]: So change it back to, if I recall correctly, just var and dict as the parameters?

File: symengine/expand.cpp (hunk 1 above)
Comment (chenchfort) [inline symengine/expand.cpp]: Can you clarify? 
`void bvisit(const Pow &self)` is already in ExpandVisitor?

File: symengine/polynomial.cpp (hunk 2 above)
Comment (isuruf) [inline symengine/polynomial.cpp]: No, I mean the deletion of line 11.

File: symengine/expand.cpp (hunk 1 above)
Comment (isuruf) [inline symengine/expand.cpp]: I meant a new method like, `pow_expand(const UnivariatePolynomial &x, unsigned long i)`
And then calling it from `void bvisit(const Pow &self)`

Comment (chenchfort) [inline symengine/expand.cpp]: I'm not familiar with expand. So not sure how you would do that.
This new pow_expand will have everything that the current `void bvisit(const Pow &self)` has?

Comment (chenchfort) [inline symengine/expand.cpp]: I'm not familiar with expand. So not sure how you would do that.
This new pow_expand will have everything that the current `void bvisit(const Pow &self)` has?

Comment (isuruf) [inline symengine/expand.cpp]: What I'm saying is adding a helper method.
`pow_expand` method contains code only for `UnivariatePolynomial`.
This method is called by `bvisit(const Pow &self)` method

Comment (chenchfort) [inline symengine/expand.cpp]: So would it essentially be this part:
`while (q != 0) {
                    if (q % 2 == 1) {
                        r = mul_uni_poly(r, p); 
                        q--;
                    }   
                    p = mul_uni_poly(p, p); 
                    q /= 2;
                }   
                _coef_dict_add_term(multiply, r); 
                return;
`

Comment (isuruf) [inline symengine/expand.cpp]: Yes

Comment (chenchfort) [inline symengine/expand.cpp]: Ok, please check if I did it correctly.

File: symengine/expand.cpp [hunk 3]
```diff
@@ -237,51 +237,99 @@ class ExpandVisitor : public BaseVisitor<ExpandVisitor> {
...
+            }   
+            x = mul_uni_poly(x, x); 
+            i /= 2;
+        }   
+        _coef_dict_add_term(multiply, r); 
+    }
+    
     void bvisit(const Pow &self) {
         RCP<const Basic> _base = expand(self.get_base());
+        if(is_a<const UnivariatePolynomial>(*self.get_base())) {
```
Comment (isuruf) [inline symengine/expand.cpp]: This check is unnecessary

File: symengine/expand.cpp [hunk 4]
```diff
@@ -237,51 +237,99 @@ class ExpandVisitor : public BaseVisitor<ExpandVisitor> {
...
-            RCP<const UnivariateIntPolynomial> p = rcp_static_cast<const UnivariateIntPolynomial>(_base);
-            RCP<const UnivariateIntPolynomial> r = univariate_int_polynomial(p->var_, {{0, integer_class(1)}});
-            while (q != 0) {
-                if (q % 2 == 1) {
-                    r = mul_poly(r, p);
-                    q--;
-                }
-                p = mul_poly(p, p);
-                q /= 2;
+            if (!is_a<Integer>(*self.get_exp()) || !is_a<Add>(*_base)) {
```
Comment (isuruf) [inline symengine/expand.cpp]: Lines 263 - 289 are unnecessary

**Comment by chenchfort** [comment]
Isuru, 

I did the modification and now one test case is failing:
/Users/Charles/symengine/symengine/tests/basic/test_polynomial.cpp:166: FAILED:
  REQUIRE( c->__str__() == "x**9 + 3*x**8 + 6*x**7 + 7*x**6 + 6*x**5 + 3*x**4 + x**3" )
with expansion:
  "(x**3 + x**2 + x)**3"
  ==
  "x**9 + 3*x**8 + 6*x**7 + 7*x**6 + 6*x**5 + 3*x**4 + x**3"

File: symengine/expand.cpp [hunk 5]
```diff
@@ -237,13 +237,57 @@ class ExpandVisitor : public BaseVisitor<ExpandVisitor> {
...
     void bvisit(const Pow &self) {
         RCP<const Basic> _base = expand(self.get_base());
+        if (is_a<Integer>(*self.get_exp()) && is_a<UnivariatePolynomial>(*_base)) {
+            unsigned long q = rcp_static_cast<const Integer>(self.get_exp())->as_int();
+            RCP<const UnivariatePolynomial> p = rcp_static_cast<const UnivariatePolynomial>(_base);
+            pow_expand(p, q);
+            return;
+        }
+
+        if (!is_a<Integer>(*self.get_exp()) || !is_a<Add>(*_base)) {
```
Comment (isuruf) [inline symengine/expand.cpp]: Lines 262 - 285 are unnecessary.

**Pull request closed without merging**